tests: integration test battery

oliver-sanders opened the pull request:
Implementation of the proposal #3616 
Closes #3547 


TODO:

- [x] Tidy some of the integration tests.
- [ ] Sort out the integration test skips.

### Highlights:

* Only ±2500 lines, nice small change (how the hell did that happen).
* Flaky unit tests now fixed.
* Successfully test runs no longer leave behind files in `~/cylc-run`.
* New integration test framework.
* Separation of CLI/scheduler functionality in `cylc/flow/scheduler*.py`.
* All three test batteries now colocated in `tests/`.

### Usage:

To run all tests:

```console
$ pytest tests/u tests/i
$ etc/bin/run-functional-tests tests/f tests/k
```

One day this will be simplified further to:

```
$ pytest tests
```

### Philosophy

There is some kind of scale with testing approaches: unit < integration < functional.

* Unit tests, quick to write, quick to debug.
* Functional tests, slow to write, slow to debug.
* (integration tests somewhere in-between).

When debugging start with the unittests and move up once you've fixed them all. The integration tests don't stand a chance of passing if the unittest are broken, the functional tests don't stand a chance of passing if the integration tests are broken.

Keep the test types separate, use the right tool for the job.

I've put README.md files in each of the test directories, take a look and see what you think, I'm not a testing expert - edits welcome!

### Integration Tests:

Documentation:

* For information on the new "integration tests" see `tests/i/README.md`
* For a functional documentation on writing integration tests see `tests/i/test_examples.py`

Implementation:

* Test flows get created in a directory hierarchy similar to the functional tests.
* These dirs get cleaned up in the event tests pass, else they are left behind for debugging.
* Don't expect to find suite logs in these directories, you wont find them!
* Captured logging is presented by pytest in traceback and can be used in tests.

Parallelism:

* The tests use `pytest.asyncio`
  * This allows you to write `async` test functions.
  * They are tested in *SERIES* just like regular pytests.
  * `pytest.asyncio` is not for parallelising tests (see GH issues).
* Pytest now uses `pytest-xdist` for parallelism:
  * Pytest is configured to run tests from the same module together.
  * This means if you define a module-scoped fixture it will only get created once.
  * So we can run a flow one, then perform multiple tests on it.

### Scheduler:

* You can now run workflows by importing the `Scheduler` class in Python.
* The CLI stuff (e.g. deamonize, `sys.exit`, logging) has been moved to `scheduler_cli`.
* You can run multiple schedulers in the same asyncio event loop.
* Interestingly this opens up running sub-suites in the same event loop as the parent.

### Caveats:

* ~Test teardown fails on NFS.~ Error suppressed until resolution of #3666
* The stdout of `cylc run` is not displayed in colour after `daemonize()`.
* The distinction between unit/integration/functional tests is a grey area.
* ~I've had to [temporally] skip the client connection tests due to an asyncio event loop incompatibility.~ Fixed.
* ~The scheduler exits with `os._exit` rather than `sys.exit`, this is not ideal.~
* ~There pytest waits for approx 10s after running integration tests, asyncio shutdown issue?~

### Requirements check-list:

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [ ] Does not contain off-topic changes (use other PRs for other changes).
   * Contains off-topic changes documented above
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.
   * documentation now in readme files under tests/

oliver-sanders commented on tests/integration/test_publisher.py:
```diff
@@ -0,0 +1,45 @@
+)
+
+
+@pytest.mark.skip('TODO: the delta doesnt seem to have an id for some reason')
```
The delta seems to be missing the id field.

oliver-sanders commented on tests/integration/test_resolvers.py:
```diff
@@ -0,0 +1,207 @@
+    assert response[0]['id'] == flow.id
+
+
+@pytest.mark.skip(
+    reason='TODO: trigger_tasks is resultin in traceback due to '
+    'missing task_globs arg')
```
Failing due to an argument error in the server?

oliver-sanders commented on tests/integration/test_data_store_mgr.py:
```diff
@@ -0,0 +1,160 @@
+    assert len(update_tasks) == len(collect_states(data, TASK_PROXIES))
+
+
+@pytest.mark.skip('TODO: fix this test')
```
The update doesn't do anything because the workflow hasn't changed state, how do I trick it?

datamel commented on cylc/flow/hostuserutil.py:
```diff
@@ -107,7 +107,7 @@ def get_local_ip_address(target):
     def get_host_ip_by_name(target):
         """Return internal IP address of target."""
-        return socket.gethostbyname(target)
+        return socket.getfqdn(target)
```
I don't think this is supposed to be here @oliver-sanders ? ;)

oliver-sanders replied:
Actually that is the right way around, the real question is how did it get the wrong way around ;)

datamel replied:
haha

oliver-sanders replied:
😳

oliver-sanders commented:
> So the idea of a functional and unit test is clear for me. The idea of functional and integration is a bit blur

Yes, the distinction is subtle and I think in truth it's really more of a scale between unit and functional rather than a distinct clustering.

I've put the framework for running suites within the integration battery to help "draw a line in the sand".

> Is that close to the right use of the new different tests

Yep, that's what I was thinking, bang on!

My thinking is:

* Unit tests to get coverage up.
* Unit/Integration tests to cover all eventualities / use cases
* One functional test to make sure that core functions are intact (but preferably only one, no need to test all eventualities here)

But I'm no testing expert! Input greatly appreciated.

oliver-sanders commented on cylc/flow/scheduler_cli.py:
```diff
@@ -241,32 +262,123 @@ def _auto_register():
+    # exit
+    LOG.info("DONE")
+    _close_logs()
+    os._exit(ret)  # sys.exit results in threading issues
```
I think it's probably ok, `sys.exit` results in `__del__` calls which enables teardown logic (which is probably what's was causing the issues).

However, I really don't like it and listed it to the caveats section in the description. There should be a better solution. The combination of daemonization and asyncio is a bit of a mind blower.

oliver-sanders replied:
Solved, FYI the trick was `loop.shutdown_asyncgens`, if only asyncio was more intuitive...

hjoliver commented:
> So the idea of a functional and unit test is clear for me. The idea of functional and integration is a bit blur ...

Someone wrote a helpful article for us :grin: 

https://www.softwaretestinghelp.com/the-difference-between-unit-integration-and-functional-testing/

kinow commented:
>I've put the framework for running suites within the integration battery to help "draw a line in the sand".

:+1 we have at least a clear definition for Cylc of what is what, and where each test should go. I can't promise I will get it right when I have to choose between the three types, but I feel confident that reading the docs and looking at existing tests I should find the right test, and eventually that will become natural :crossed_fingers: 

> > So the idea of a functional and unit test is clear for me. The idea of functional and integration is a bit blur ...
> 
> Someone wrote a helpful article for us grin
> 
> https://www.softwaretestinghelp.com/the-difference-between-unit-integration-and-functional-testing/

Yup! I had seen this one too. And each directory has the helpful `README` that @oliver-sanders wrote. If anyone has problems or is confused about which one to write, I think at least someone can point to those files and also to existing tests as reference.

oliver-sanders commented:
Yep, I think I used the diagram from that article in my proposal. I did a bit of reading which was quite confidence giving seeing people confirm what I've been slowly realising.

The really big bit is time to write and time to debug. We spend so much time on functional tests, I personally sunk nearly an entire week into them (less than the time it took to write this new framework) just to fix broken tests. Hopefully this will help accelerate future development, might take a while to get there though...

oliver-sanders commented:
Note, I've found that `pytest` doesn't like symlinks much so `pytest tests/` will actually run each test battery twice (once for the directory, once for the shorthand symlink).

Looking into it...

oliver-sanders commented:
Rebased and deconflicted, test changes ported from #3500.

oliver-sanders commented on tests/integration/conftest.py:
```diff
@@ -0,0 +1,243 @@
+    ))
+
+
+def _rmtree(path, tries=5):
+    """Remove a directory and its contents.
+
+    Attempt to remove a directory up to ``tries`` times before failing.
+
+    This should be used in-place of ``shutil.rmtree`` which doesn't
+    work reliably on NFS.
+
+    See also:
+        * https://stackoverflow.com/questions/58943374/
+          shutil-rmtree-error-when-trying-to-remove-nfs-mounted-directory
+        * https://bugzilla.redhat.com/show_bug.cgi?id=1362667
+
+    Args:
+        path (pathlib.Path): The directory to remove.
+        tries (int): Maximum number of attempts to make.
+
+    """
+    exc = None
+    for _ in range(tries):
+        try:
+            rmtree(path)
+            return
+        except OSError as exc:
+            # give NFS some thinking time
+            sleep(0.01)
+    raise exc
```
This fixes a bug spotted by @datamel where `shutul.rmtree` fails on NFS.

It's a nasty hack but it does the job, here are some examples of how other projects have handled this issue:

* https://github.com/hashdist/hashdist/pull/116/files#diff-f8e43180100ba3086d46440b72e3e8f3
* https://github.com/davidemms/OrthoFinder/commit/3d56a6e4870058f78721fa733ae27f60ab4d0c21#diff-4b6271e9a0c35f1d832125bcadf25034

datamel commented on tests/integration/conftest.py:
```diff
@@ -72,6 +73,35 @@ def _pytest_passed(request):
+        except OSError as exc:
+            # give NFS some thinking time
+            sleep(0.01)
+    raise exc
```
I am getting scope problems with this 
![image](https://user-images.githubusercontent.com/37735232/85297523-c0c01b00-b49a-11ea-8fd4-4647e338e6a0.png)
Did you mean to tab that raise into the except?
```suggestion
            raise exc
```

oliver-sanders replied:
Whaaa, but it's defined on line 94!

If you indent the `raise exc` then it will only try once rather than multiple times.

oliver-sanders replied:
Curiously if you do:

```
except OSError as exc_:
    exc = exc_
```

It's fine, I think I've just learned something new about Python3 scoping which was unexpected.

The NFS issue sadly isn't fixed, I think that asyncio might be keeping the file-descriptors alive which will cause NFS to write those pesky `.nfs` files you can't get rid of until the process has died, since the process is pytest and pytest is waiting for NFS its a deadlock.

The solution might coincide with the `os._exit` problem if lucky.

datamel replied:
I've had a good look at this today. I have tried making use of the rmtree callback function onerror to fix it but so far with no success. There is an interesting explanation here...  https://stackoverflow.com/questions/11228079/python-remove-directory-error-file-exists
So, by that response, if we can close the file in cylc, we shouldn't have the problem.
Looking at the .nfs files created, they look like they're garbled database files...
![image](https://user-images.githubusercontent.com/37735232/85343781-ae1f0380-b4e5-11ea-9daa-26c852e33649.png)
Might help us locate where the offending files are being created and then we can properly close them to avoid the issue. 
Incidentally, when running the integration tests, I'm also getting errors of...
![image](https://user-images.githubusercontent.com/37735232/85343706-744dfd00-b4e5-11ea-8568-3b932aeec60f.png)

datamel replied:
Ah, I've just re-read your message. 
> since the process is pytest and pytest is waiting for NFS its a deadlock.
I had asssumed the process wasn't pytest as I still can not delete the files manually after pytest has completed its run. This is a pretty frustrating problem!

datamel replied:
Ok, I have a fix (if temporary until a better one comes along!)...  
`rmtree(path, ignore_errors=True)` 
This will still leave the directories in place if .nfs files are there but at least the tests pass and we can manually delete them whenever we are able?

oliver-sanders replied:
I'm now fairly confident that asyncio is what's keeping the file descriptors open so I'll try and fix that underlying issue which in so-doing will strike off a couple of items under **caveats** in the description.

This needn't delay review but you may want to apply your patch or else just comment out the `_rmtree` bit for testing in the mean time.

oliver-sanders replied:
Whilst we're here, if you see the `n_tries` error, it comes from calling `schd.shutdown` from outside of the `main_loop`. At the moment `shutdown` must be in lock-step with the `main_loop` (else the main loop tries to do things whilst `shutdown` is busy tearing stuff apart).

The NFS/asyncio fix will also fix this 😀.

datamel replied:
> Whilst we're here, if you see the n_tries error, it comes from calling schd.shutdown from outside of the main_loop. 

Ah good to know :)

oliver-sanders commented:
Rebased and de-conflicted, this PR is going to be horrible to rebase!

@datamel you might want to check I've done your auth change proud with a quick `git diff master oliver/scheduler-options` as the code has been bumped around a bit.

hjoliver commented:
So I started having a good look at this today.  Like @kinow it will take a bit of learning next time I write a test, but generally it looks like a great improvement.  Is it going to be easy, do you think, to figure out exactly what can be moved out of existing functional tests and into integration tests?

datamel commented on tests/integration/conftest.py:
```diff
@@ -72,6 +72,33 @@ def _pytest_passed(request):
+        path (pathlib.Path): The directory to remove.
+        tries (int): Maximum number of attempts to make.
+
+    """
+    for _ in range(tries):
+        try:
+            rmtree(path)
+            return
+        except OSError as exc:
+            pass
+    raise exc
```
Still getting scope problems with this... think this fixes it... 
```suggestion
    """
    outer_exc = None
    for _ in range(tries):
        try:
            rmtree(path)
            return
        except OSError as exc:
            outer_exc = exc
            pass
    raise outer_exc
```

oliver-sanders commented:
>  Is it going to be easy, do you think, to figure out exactly what can be moved out of existing functional tests and into integration tests?

Not too hard if you know what the intention of the test was in the first place. It's probably easier to work out what tests should be functional tests than which ones should be ported:

If it's not intending to test:
 
* Workflow execution.
* Shell interaction.
* System integration.
* Job execution.

Then it's either a unit or integration test.

There's one piece of the puzzle missing from the integration test battery at the moment which is the ability to mockup the db so you can (re)start the flow in a pre-determined state. (we'll have to save this beauty for a later PR).

Once that's done quite a few types of test can be ported:

* Tests which put a flow in a strange state then call server endpoints.
* CLI tests can be converted into server endpoint tests (above).
* Tests which put a flow in a strange state to see what happens next.
* Workflow scanning.

hjoliver commented:
(Nicely documented in the README files, BTW).

oliver-sanders commented:
* Rebased.
* Deconflicted.
* Fixed asyncio issues (you will need to `pip install` again).
* Suppressed NFS issue as [unrelated to this change](https://github.com/cylc/cylc-flow/issues/3666).

oliver-sanders commented:
@datamel and @kinow thank you for stepping in to review.

There are a couple of specific aspects which could do with standalone reviewing:

* @hjoliver best make sure you're happy with the scheduler / scheduler_cli re-alignment and the general gist of the change.
* @dwsutherland if you could cast your eye over the asyncio stuff in scheduler/scheduler_cli.
* @dwsutherland best make sure I've not butchered the tests.

datamel commented on pytest.ini:
```diff
@@ -17,13 +17,14 @@
 testpaths =
-    cylc/flow/
-    tests/lib/python/
+    tests/unit/
```
```suggestion
    tests/unit/
    tests/integration/
```

oliver-sanders replied:
I purposefully left the integration tests out of `testpaths` so that when you run `pytest` it defaults to running `tests/unit`.

datamel replied:
Ah, no worries :)

oliver-sanders commented:
The last commit should fix the hanging functional test. Hopefully this is the last time I have to mess with asyncio in this PR!

Even though it's only a few lines getting the asyncio code right has been a lot of work, the integration test framework only took one day. Here are some lessons learned:

* I no longer believe in magic, but I now believe in ghosts.
* Async generators must be manually shutdown, you might not use async generators in your code but that doesn't mean they aren't there. If you use `asyncio.run` it creates and closes an event loop for you tidying up async gens correctly. Look at the source code for `asyncio.run`, it's not actually doing much under-the-hood.
* If `sys.exit` hangs it means you've got threads (which have `sleep` statements) that haven't been shutdown correctly. Keep track of threads closely and ensure they are always shutdown. If `os._exit` fixes it then this is probably the issue.
* Daemonisation is a hard barrier that event loops cannot persist through. This is totally expected but it means you need to tidy up *before* daemonisation else the original process won't be able to exit.

oliver-sanders commented:
Merged and de-conflicted (tried rebase, it was horrible).

oliver-sanders commented:
Fixed `tests/f/cylc-ping/03` which was dependent on `key_housekeeping` being called before `remrun`.

hjoliver commented:
> Here are some lessons learned:...

That sounds like some stuff that I don't want to rediscover the hard way in the future. Would it be worth copying to code comments for reference (in `scheduler.py`?).

oliver-sanders commented:
Yep, will do.

oliver-sanders commented:
Deconflicted.

oliver-sanders commented on cylc/flow/scheduler.py:
```diff
@@ -561,17 +490,159 @@ def configure(self):
+            await self.initialise()
+            await self.configure()
+            await self.start_servers()
+            await self.log_start()
```
> so I guess you did this in anticipation of better use of asyncio in the future?

Actually some of them do await things, so I thought it would be better to use an await interface for all of them to avoid having to individually upgrade them later.

oliver-sanders commented on pytest.ini:
```diff
@@ -17,13 +17,15 @@
-    --ignore=cylc/flow/data_messages_pb2.py
+    --ignore=tests/unit/parsec/getcfg/bin/one-line.py
+    --ignore=tests/unit/parsec/nullcfg/bin/empty.py
+    --ignore=tests/unit/parsec/synonyms/bin/synonyms.py
```
Will remove.
